core/sdk/src/main/java/com/mongodb/stitch/core/auth/providers/userapikey/CoreUserApiKeyAuthProviderClient.java

```diff
@@ -16,10 +16,142 @@
 
-public abstract class CoreUserApiKeyAuthProviderClient {
-  private final String providerName;
+import com.mongodb.stitch.core.auth.internal.StitchAuthRequestClient;
+import com.mongodb.stitch.core.auth.internal.StitchAuthRoutes;
+import com.mongodb.stitch.core.auth.providers.CoreAuthProviderClient;
+import com.mongodb.stitch.core.auth.providers.userapikey.models.UserApiKey;
+import com.mongodb.stitch.core.internal.net.Method;
+import com.mongodb.stitch.core.internal.net.StitchAuthDocRequest;
+import com.mongodb.stitch.core.internal.net.StitchAuthRequest;
 
-  protected CoreUserApiKeyAuthProviderClient(final String providerName) {
-    this.providerName = providerName;
+import java.util.List;
+
+import org.bson.Document;
+import org.bson.types.ObjectId;
+
+public class CoreUserApiKeyAuthProviderClient
+        extends CoreAuthProviderClient<StitchAuthRequestClient> {
+  private final Routes routes;
+
+  protected CoreUserApiKeyAuthProviderClient(final StitchAuthRequestClient authRequestClient,
+                                             final StitchAuthRoutes authRoutes) {
+    super(UserApiKeyAuthProvider.DEFAULT_NAME, authRequestClient,
+            authRoutes.getBaseAuthRoute() + "/api_keys");
+    this.routes = new Routes(getBaseRoute());
+  }
+
+  /**
+   * Creates a user API key that can be used to authenticate as the current user.
+   *
+   * @param name The name of the API key to be created
+   */
+  protected UserApiKey createApiKeyInternal(final String name) {
+    final StitchAuthDocRequest.Builder reqBuilder = new StitchAuthDocRequest.Builder();
+    reqBuilder
+            .withMethod(Method.POST)
+            .withPath(this.getBaseRoute())
+            .withDocument(new Document(Routes.ApiKeyFields.NAME, name))
+            .withRefreshToken()
+            .withShouldRefreshOnFailure(false);
+    return decode(getRequestClient().doAuthenticatedRequest(reqBuilder.build()), UserApiKey.class);
+  }
+
+  /**
+   * Fetches a user API key associated with the current user.
+   *
+   * @param id The id of the API key to be fetched.
+   */
+  protected UserApiKey fetchApiKeyInternal(final ObjectId id) {
+    final StitchAuthRequest.Builder reqBuilder = new StitchAuthRequest.Builder();
+    reqBuilder
+            .withMethod(Method.GET)
+            .withPath(routes.getApiKeyRouteForId(id.toHexString()))
+            .withRefreshToken()
+            .withShouldRefreshOnFailure(false);
+    return decode(getRequestClient().doAuthenticatedRequest(reqBuilder.build()), UserApiKey.class);
+  }
+
+  /**
+   * Fetches the user API keys associated with the current user.
+   */
+  protected List<UserApiKey> fetchApiKeysInternal() {
+    final StitchAuthRequest.Builder reqBuilder = new StitchAuthRequest.Builder();
+    reqBuilder
+            .withMethod(Method.GET)
+            .withPath(this.getBaseRoute())
+            .withRefreshToken()
+            .withShouldRefreshOnFailure(false);
+    return decodeList(
```

Comment:
If you add the ability to do decoding with `Decoder<T>` to doAuthenticatedRequest like the Json variants, you can just use this which is in my current branch:
```
class CollectionDecoder<TResult> implements Decoder<Collection<TResult>> {

  private final Decoder<TResult> decoder;

  public CollectionDecoder(final Decoder<TResult> decoder) {
    this.decoder = decoder;
  }

  @Override
  public Collection<TResult> decode(final BsonReader reader, final DecoderContext decoderContext) {
    final Collection<TResult> docs = new ArrayList<>();
    reader.readStartArray();
    while (reader.readBsonType() != BsonType.END_OF_DOCUMENT) {
      final TResult doc = decoder.decode(reader, decoderContext);
      docs.add(doc);
    }
    reader.readEndArray();
    return docs;
  }
}
```

e.g.
```
return service.callFunctionInternal(
        "find",
        Collections.singletonList(args),
        new CollectionDecoder<>(decoder));
```